### PR TITLE
FIR/UAST: commonize type-related expressions

### PR DESCRIPTION
**Pull request by jsjeon**
1st commit adds an API to FIR IDE that converts `coneType` retrieved from `KtExpression` to `PsiType` and uses it in FIR UAST.

Before commonizing type reference expression, 2nd commit unifies the use of that in FE1.0 UAST so that every creation is using a lazy version. Then, 3rd commit commonizes type reference expression.

4th commit adds an API to FIR IDE that converts `coneType` retreived from `KtTypeReference` to `PsiType` and uses it in FIR UAST.

5th commit is a mechanical migration of kinds constants.

Since then, remaining commits are commonizing type-related expressions:
* `is`, represented as type check expression

* `as`, represented as binary expression with type; other binary expression in general needs operator resolution

* array access; can't see any test changes from legacy test data, so added before this change. Also, not fully commonized since FE1.0 UAST (or `BindingContext`) needs a special handling to retrieve array access type from array type (while FIR counterpart can reuse the expression type retrieval from the base service

------

Last commits address review comments:
* UAST-specific version of `analyse` HL API
* `KtPsiTypeProvider` in HL API whose name implies that it is `PsiType`-specific component/provider

### CODE REVIEW

**Comment by jsjeon** [comment]
Will keep this as a draft until #4428 is merged.

File: idea/idea-frontend-fir/src/org/jetbrains/kotlin/idea/frontend/api/fir/components/KtFirExpressionTypeProvider.kt [hunk 1]
```diff
@@ -50,6 +52,18 @@ internal class KtFirExpressionTypeProvider(
         }
     }
 
+    override fun getPsiTypeForKtExpression(
```
Comment (darthorimar) [inline idea/idea-frontend-fir/src/org/jetbrains/kotlin/idea/frontend/api/fir/components/KtFirExpressionTypeProvider.kt:55]: Let's extract `getPsiTypeForKtExpression` and `getPsiType(ktTypeReference: KtTypeReference..)` into separate component, smth like `KtPsiTypeProvider` and move all firType -> psiType mapping from light classes to it. And just use those new functions from light classes. This way we get rid of cyclic dependency between light classes and HL API

**Review by darthorimar — CHANGES_REQUESTED** [review]

File: plugins/uast-kotlin-fir/src/org/jetbrains/uast/kotlin/FirKotlinUastResolveProviderService.kt [hunk 2]
```diff
@@ -36,8 +37,10 @@ interface FirKotlinUastResolveProviderService : BaseKotlinUastResolveProviderSer
     }
 
     override fun getExpressionType(uExpression: UExpression): PsiType? {
-        // TODO("Not yet implemented")
-        return null
+        val ktExpression = uExpression.sourcePsi as? KtExpression ?: return null
+        analyseWithCustomToken(ktExpression, AlwaysAccessibleValidityTokenFactory) {
```
[the remark this answers is not in the document]
Comment (jsjeon) [reply plugins/uast-kotlin-fir/src/org/jetbrains/uast/kotlin/FirKotlinUastResolveProviderService.kt]: Sounds good. Added in the 2nd-to-last commit.

File: idea/idea-frontend-fir/src/org/jetbrains/kotlin/idea/frontend/api/fir/components/KtFirExpressionTypeProvider.kt (hunk 1 above)
Comment (jsjeon) [reply idea/idea-frontend-fir/src/org/jetbrains/kotlin/idea/frontend/api/fir/components/KtFirExpressionTypeProvider.kt]: Extracted those utils to `KtPsiTypeProvider` as per the suggestion. See the last commit.

Re: cyclic dependecy,

actually, `asPsiType` is inside the same module, i.e., not in light classes, but in idea-frontend-fir. Simply it was my mistake to open that as `public`; `internal` should be good enough, and changed that way.

**Review by xiexed — APPROVED** [review]

**Comment by darthorimar** [comment]
Merged manually, thank you :)

**Pull request closed without merging**